### PR TITLE
RavenDB-19621 : Sharding - Block rolling indexes

### PR DESCRIPTION
**Pull request by aviv86**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19621

### Additional description

- modify `AbstractIndexCreateController` and `ShardedIndexCreateController` to throw `NotSupportedInShardingException` on attempt to deploy rolling index in sharded database
- change all the index controllers types in `IndexStore` to their abstract types, use the regular database index controllers for non-sharded db and the sharded index controllers for a sharded db
-  this change was made in order to reach `ShardedIndexCreateController` (instead of the non-sharded version) when creating an auto index via index store ([here](https://github.com/ravendb/ravendb/blob/v6.0/src/Raven.Server/Documents/Queries/Dynamic/DynamicQueryRunner.cs#L170))

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- Need to document that rolling indexes are blocked in sharded database 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

**Review by karmeli87 — APPROVED** [review]
LGTM

File: src/Raven.Server/Documents/Indexes/IndexStore.cs [hunk 1]
```diff
@@ -73,36 +73,40 @@ public class IndexStore : IDisposable
...
 
             var stoppedConcurrentIndexBatches = _documentDatabase.Configuration.Indexing.NumberOfConcurrentStoppedBatchesIfRunningLowOnMemory;
             StoppedConcurrentIndexBatches = new SemaphoreSlim(stoppedConcurrentIndexBatches);
         }
 
+        public IndexStore(DocumentDatabase documentDatabase, ServerStore serverStore) : this(documentDatabase)
+        {
+            ServerStore = serverStore;
+            Create = new DatabaseIndexCreateController(documentDatabase);
+        }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Indexes/IndexStore.cs]: I think we should have here a single `ctor`

File: src/Raven.Server/Documents/Indexes/IndexStore.cs [hunk 2]
```diff
@@ -73,36 +73,40 @@ public class IndexStore : IDisposable
...
+        private bool _initialized;
 
-        public readonly DatabaseIndexDeleteController Delete;
+        private bool _run = true;
 
-        public readonly DatabaseIndexCreateController Create;
+        private long _lastSurpassedAutoIndexesDatabaseRecordEtag;
 
-        public readonly DatabaseIndexHasChangedController HasChanged;
+        private const int PathLengthLimit = 259; // Roslyn's MetadataWriter.PathLengthLimit = 259
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Indexes/IndexStore.cs]: Can we revert all the place shifting above, it seems to be not necessary.

File: src/Raven.Server/Documents/Indexes/Sharding/ShardedIndexStore.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using Raven.Server.ServerWide;
+
+namespace Raven.Server.Documents.Indexes.Sharding
+{
+    public class ShardedIndexStore : IndexStore
+    {
+        public ShardedIndexStore(DocumentDatabase database, ServerStore serverStore) : base(database)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Indexes/Sharding/ShardedIndexStore.cs]: let's pass `ShardedDocumentDatabase` here

File: src/Raven.Server/Documents/Indexes/Sharding/ShardedDatabaseIndexCreateController.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+﻿using JetBrains.Annotations;
+using Raven.Client.Documents.Indexes;
+using Raven.Client.Exceptions.Sharding;
+
+namespace Raven.Server.Documents.Indexes.Sharding;
+
+public class ShardedDatabaseIndexCreateController : DatabaseIndexCreateController
+{
+    public ShardedDatabaseIndexCreateController([NotNull] DocumentDatabase database)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Indexes/Sharding/ShardedDatabaseIndexCreateController.cs]: let's pass ShardedDocumentDatabase here

File: src/Raven.Server/Documents/Sharding/ShardedDocumentDatabase.cs [hunk 5]
```diff
@@ -16,8 +18,8 @@ namespace Raven.Server.Documents.Sharding;
 public class ShardedDocumentDatabase : DocumentDatabase
 {
     public readonly int ShardNumber;
-
-    public readonly string ShardedDatabaseName;
+    
+    public string ShardedDatabaseName;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/ShardedDocumentDatabase.cs]: why this change?

File: src/Raven.Server/Documents/Indexes/Sharding/ShardedDatabaseIndexCreateController.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        : base(database)
+    {
+    }
+
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Indexes/Sharding/ShardedDatabaseIndexCreateController.cs:14]: do we need to override the `ValidateStaticIndexAsync` as well?

**Review by karmeli87 — COMMENTED** [review]
LGTM

File: src/Raven.Server/Documents/Indexes/Sharding/ShardedDatabaseIndexCreateController.cs (hunk 6 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Indexes/Sharding/ShardedDatabaseIndexCreateController.cs:14]: don't thinks it's needed, static indexes are handled by `ShardedIndexCreateController`

File: src/Raven.Server/Documents/Sharding/ShardedDocumentDatabase.cs (hunk 5 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Sharding/ShardedDocumentDatabase.cs]: reverted

File: src/Raven.Server/Documents/Indexes/Sharding/ShardedDatabaseIndexCreateController.cs (hunk 4 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Indexes/Sharding/ShardedDatabaseIndexCreateController.cs]: done

File: src/Raven.Server/Documents/Indexes/Sharding/ShardedIndexStore.cs (hunk 3 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Indexes/Sharding/ShardedIndexStore.cs]: done

File: src/Raven.Server/Documents/Indexes/IndexStore.cs (hunk 2 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Indexes/IndexStore.cs]: fixed

File: src/Raven.Server/Documents/Indexes/IndexStore.cs (hunk 1 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Indexes/IndexStore.cs]: done

**Review by karmeli87 — APPROVED** [review]
